Support controllers_path generator option for controller and manifest tasks.

jhirn opened the pull request:
This PR aims to support paths other than `app/javascript` for the root path in generator and manifest tasks. [Similar](https://github.com/hotwired/stimulus-rails/pull/119) [attempts](https://github.com/hotwired/stimulus-rails/issues/148) have been made/requested, but this PR uses `config.generators` instead of an initializer to set `controllers_path`.  A small drawback is the manifest tasks load `:environment` to access configuration from `application.rb`, but it plays nicely with generator help message and doesn't pollute the `Stimulus` namespace. Here's the usage:

```ruby
module Rails4Lyfe
  class Application < Rails::Application
    ...

    config.generators do |g|
      ...
      g.stimulus :stimulus, skip_manifest: true, controllers_path: 'app/frontend/controllers'
      ...
    end
  end
end
```

I like using the generator because it keeps the controller consistent with the `// Connects to data-controller=..` comment and naming conventions which get tricky with camelCase, under_score and dash-erized in play. Because the [path is hard coded](https://github.com/hotwired/stimulus-rails/blob/96ae4e835af18d4f7eacc73fc7bfa1267bff0bcb/lib/generators/stimulus/stimulus_generator.rb#L10) in the generator, the only option is to monkey patch it. Monkey patching the generator in my project made it difficult to reference the template from inside the gem requiring me to copy template as well. 

If there's support for this option I'd be happy to proceed with updating docs and tests, although I had a pretty tough time getting the suite to run locally.

Thank you for considering this PR.

jhirn committed:
Support js_root generator option for controller and manifest tasks.

dhh commented on lib/tasks/stimulus_tasks.rake:
```diff
@@ -7,6 +7,11 @@ module Stimulus
     end
 
+    def controllers_path
+      js_root = Rails.application.config.generators.options.dig(:stimulus, :js_root) || "app/javascript"
```
Wouldn't js_root always be set, given that you have a default for the class_option? If so, I don't think we need to repeat the default here. And in fact, I'd just inline the entire call in the Rails.root.join.

jhirn replied:
I pushed up this change. LMK what you think. 

However this pans out thanks for the code review 👍

dhh commented on lib/generators/stimulus/stimulus_generator.rb:
```diff
@@ -4,10 +4,11 @@ class StimulusGenerator < Rails::Generators::NamedBase # :nodoc:
   source_root File.expand_path("templates", __dir__)
 
   class_option :skip_manifest, type: :boolean, default: false, desc: "Don't update the stimulus manifest"
+  class_option :js_root, type: :string, default: 'app/javascript', desc: "Root path for javascript files"
```
Always use " instead of ' for strings.

jhirn committed:
Use StimulusGenerator .class_options for js_root in rake task.

jhirn committed:
Use .js_root instead place of hash accessor

dhh commented on lib/tasks/stimulus_tasks.rake:
```diff
@@ -46,16 +55,16 @@ namespace :stimulus do
+
     task :display do
-      puts Stimulus::Manifest.generate_from(Rails.root.join("app/javascript/controllers"))
+      Stimulus::Manifest.generate_from(Stimulus::Tasks.controllers_path)
```
Why remove the puts here? That would remove the actual display of the Manifest.

jhirn replied:
Gladly.  It was nice when I was debugging, but I'm usually beaten with a stick for leaving `puts` in code. 😅

dhh replied:
puts for debugging is definitely not something you want to ship, but this puts is in service of what the task does: Display the manifest to the user on the terminal.

jhirn replied:
@dhh What are your thoughts adding a puts to `stimulus:manifest:update` to avoid surprise when `js_root` is not default?

```
desc "Update the Stimulus manifest (will overwrite controllers/index.js)"
    task :update do
      puts "Updating stimulus manifest in #{Stimulus::Tasks.controllers_path.relative_path_from(Rails.root)}"
      manifest = Stimulus::Manifest.generate_from(Stimulus::Tasks.controllers_path)
       ....
```

dhh commented on lib/tasks/stimulus_tasks.rake:
```diff
@@ -7,6 +8,14 @@ module Stimulus
+      Rails.root.join(StimulusGenerator.new(["stimulus"], generator_options, {}).options.js_root, "controllers")
+    end
+
+    def generator_options
```
Since this isn't called anywhere else, I'd just inline it in `controllers_path`.

dhh replied:
I like:

```ruby
def controllers_path
  generator = StimulusGenerator.new(["stimulus"], Rails.application.config.generators.options.dig(:stimulus), {}) 
  Rails.root.join generator.options.js_root, "controllers"
end
```

jhirn committed:
Inline generator_options as they aren't used elsewhere

jhirn committed:
add puts back to stimulus:manifest:display

dhh committed:
Style changes

dhh committed:
Use a fully formed controllers_path

dhh committed:
Use controllers_path

dhh committed:
Style

dhh commented:
Made some changes. Think we probably need to add a test for this, though.

jhirn commented:
Thanks for the updates.  I've updated the title and original post to reference `controllers_path` in favor of `js_root`. I can add some tests as well. 

How can I avoid  `uninitialized constant ActiveSupport::LoggerThreadSafeLevel::Logger (NameError)` when running `bundle exec rake test`? I installed Ruby 3.1.7 and added `require "logger"` to `test_helper.rb` which gets me running but, I probably shouldn't have to and still get a warning: 

```
/Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.10/lib/active_support/logger_thread_safe_level.rb:16:in `<module:LoggerThreadSafeLevel>': uninitialized constant ActiveSupport::LoggerThreadSafeLevel::Logger (NameError)

    Logger::Severity.constants.each do |severity|
    ^^^^^^
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.10/lib/active_support/logger_thread_safe_level.rb:9:in `<module:ActiveSupport>'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.10/lib/active_support/logger_thread_safe_level.rb:8:in `<top (required)>'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.10/lib/active_support/logger_silence.rb:5:in `require'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.10/lib/active_support/logger_silence.rb:5:in `<top (required)>'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.10/lib/active_support/logger.rb:3:in `require'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.10/lib/active_support/logger.rb:3:in `<top (required)>'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.10/lib/active_support.rb:29:in `require'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/activesupport-6.1.7.10/lib/active_support.rb:29:in `<top (required)>'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/railties-6.1.7.10/lib/rails/command.rb:3:in `require'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/railties-6.1.7.10/lib/rails/command.rb:3:in `<top (required)>'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/railties-6.1.7.10/lib/rails/cli.rb:12:in `require'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/railties-6.1.7.10/lib/rails/cli.rb:12:in `<top (required)>'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/railties-6.1.7.10/exe/rails:10:in `require'
        from /Users/jhirn/.rbenv/versions/3.1.6/lib/ruby/gems/3.1.0/gems/railties-6.1.7.10/exe/rails:10:in `<top (required)>'
        from /Users/jhirn/.rbenv/versions/3.1.6/bin/rails:25:in `load'
        from /Users/jhirn/.rbenv/versions/3.1.6/bin/rails:25:in `<main>'
.
```

Two follow up questions: 

* Should we add something to the readme? I didn't see a section for generator configuration otherwise I'd amend it. 
* Should we observe `controllers_path` for `stimulus:install:*` tasks?

dhh commented:
I've fixed the test issue. Merge with main, and then you can run bin/test.

Yes, we should document this.

Not sure what you mean about observing `controllers_path`?

jhirn commented:
> Not sure what you mean about observing `controllers_path`?

The templates use by `rails stimulus:install` are hard coded and do not respect `controllers_path`. Should they?

jhirn commented:
Got side tracked with life but I should have some tests up by before next week.

jhirn committed:
Fix rebase: controller_path => controller_root

jhirn committed:
Add tests. Validate controller_path is set and not absolute directory.

jhirn commented:
@dhh Pushed up some tests and added validations for blank and absolute controller path ( blank path becomes absolute because we concatenate with `/#{controller_name}`. Both cases would fail with ` Read-only file system @ rb_sysopen` so I added specific RuntimeError messages to be a bit more friendly. 

I could also see using default when blank and converting absolute paths to relative by removing any leading file system separators. It's a trade off between strictness and ergonomics. I personally prefer the latter. 

As far as documentation, I only updated USAGE to give an example of controllers-path. I feel like the README should have a section detailing generator configuration that includes `skip_manifest` in addition to `controllers_path` but didn't want to overstep. Not sure if that should be part of this or a separate PR to fully document generator configuration.

tomu123 commented on lib/tasks/stimulus_tasks.rake:
```diff
@@ -47,12 +53,12 @@ namespace :stimulus do
     desc "Update the Stimulus manifest (will overwrite controllers/index.js)"
-    task :update do
-      Stimulus::Manifest.write_index_from(Rails.root.join("app/javascript/controllers"))
+    task update: :environment do
```
Why do we need to add the environment task as a dependency? It doesn't seem necessary.

tomu123 commented on lib/tasks/stimulus_tasks.rake:
```diff
@@ -7,6 +8,11 @@ module Stimulus
     end
 
+    def controllers_root
+      Rails.root.join \
```
It's a little hard to read and still too long in my opinion, I believe the suggestion from @dhh should be good
```
def controllers_path
  generator = StimulusGenerator.new(["stimulus"], Rails.application.config.generators.options.dig(:stimulus), {}) 
  Rails.root.join generator.options.js_root, "controllers"
end
```

tomu123 commented:
amazing work @jhirn, I really would like to see this merged since i have a use case for this in one of the app I am currently working on. Do you know when it may be merged, @dhh?
